src/Definition/ObjectDefinition.php

```diff
@@ -223,6 +224,25 @@ public function replaceNestedDefinitions(callable $replacer)
         });
     }
 
+    /**
+     * Replaces all the wildcards in the string with the given replacements.
+     *
+     * @param string[] $replacements
+     */
+    public function replaceWildcards(array $replacements)
+    {
+        $className = $this->getClassName();
+
+        foreach ($replacements as $replacement) {
+            $pos = strpos($className, DefinitionArray::WILDCARD);
+            if ($pos !== false) {
+                $className = substr_replace($className, $replacement, $pos, 1);
+            }
+        }
+
+        $this->setClassName($className);
+    }
```

Comment:
I've moved the logic of replacing wildcards into this class, that allows for unit tests and a simpler code in `DefinitionArray`. This is not directly related to your bugfix but it improves the code a bit I think.